File: DSA/Java/test/dsa/lib/utils/range/StripTest.java
package dsa.lib.utils.range;

import dsa.lib.utils.Range;
import org.junit.Test;

import static org.junit.Assert.assertArrayEquals;

public class StripTest {
    public static class LStripTest {
        private int[] lstripIt(int[] nums) {
            return Range.lstripIt(nums);
        }

        private int[] lstripDfs(int[] nums) {
            return Range.lstripDfs(nums);
        }

        @Test
        public void empty_returnEmpty() {
            assertLStrip(new int[]{}, new int[]{});
        }

        private void assertLStrip(int[] expected, int[] nums) {
            assertArrayEquals(expected, lstripIt(nums));
            assertArrayEquals(expected, lstripDfs(nums));
        }

        @Test
        public void all0s_returnEmpty() {
            assertLStrip(new int[]{}, new int[]{0});
            assertLStrip(new int[]{}, new int[]{0, 0});
        }

        @Test
        public void leftNot0s() {
            assertLStrip(new int[]{1}, new int[]{1});
            assertLStrip(new int[]{1, 2}, new int[]{1, 2});
        }

        @Test
        public void left0s() {
            assertLStrip(new int[]{1}, new int[]{0, 1});
            assertLStrip(new int[]{1}, new int[]{0, 0, 1});
            assertLStrip(new int[]{1, 2}, new int[]{0, 0, 1, 2});
            assertLStrip(new int[]{1, 0, 2}, new int[]{0, 0, 1, 0, 2});
        }
    }

    public static class RStripTest {
        private int[] rstripIt(int[] nums) {
            return Range.rstripIt(nums);
        }

        private int[] rstripDfs(int[] nums) {
            return Range.rstripDfs(nums);
        }

        private void assertRStrip(int[] expected, int[] nums) {
            assertArrayEquals(expected, rstripIt(nums));
            assertArrayEquals(expected, rstripDfs(nums));
        }

        @Test
        public void empty_returnEmpty() {
            assertRStrip(new int[]{}, new int[]{});
        }

        @Test
        public void all0s_returnEmpty() {
            assertRStrip(new int[]{}, new int[]{0});
            assertRStrip(new int[]{}, new int[]{0, 0});
        }

        @Test
        public void rightNot0s() {
            assertRStrip(new int[]{1}, new int[]{1});
            assertRStrip(new int[]{1, 2}, new int[]{1, 2});
        }

        @Test
        public void right0s() {
            assertRStrip(new int[]{1}, new int[]{1, 0});
            assertRStrip(new int[]{1}, new int[]{1, 0, 0});
            assertRStrip(new int[]{1, 2}, new int[]{1, 2, 0, 0});
            assertRStrip(new int[]{1, 0, 2}, new int[]{1, 0, 2, 0, 0});
        }

    }

    public static class _StripTest {
        private int[] stripIt(int[] nums) {
            return Range.stripIt(nums);
        }

        private int[] stripDfs(int[] nums) {
            return Range.stripDfs(nums);
        }

        private void assertStrip(int[] expected, int[] nums) {
            assertArrayEquals(expected, stripIt(nums));
            assertArrayEquals(expected, stripDfs(nums));
        }

        @Test
        public void empty_returnEmpty() {
            assertStrip(new int[]{}, new int[]{});
        }

        @Test
        public void all0s_returnEmpty() {
            assertStrip(new int[]{}, new int[]{0});
            assertStrip(new int[]{}, new int[]{0, 0});
        }

        @Test
        public void leftRightNot0s() {
            assertStrip(new int[]{1}, new int[]{1});
            assertStrip(new int[]{1, 2}, new int[]{1, 2});
        }

        @Test
        public void leftRight0s() {
            assertStrip(new int[]{1}, new int[]{0, 1, 0});
            assertStrip(new int[]{1}, new int[]{0, 0, 1, 0});
            assertStrip(new int[]{1, 2}, new int[]{0, 0, 1, 2});
            assertStrip(new int[]{1, 2}, new int[]{0, 0, 1, 2, 0, 0});
            assertStrip(new int[]{1, 0, 2}, new int[]{0, 0, 1, 0, 2});
            assertStrip(new int[]{1, 0, 2}, new int[]{0, 0, 1, 0, 2, 0, 0});
        }
    }
}

File: DSA/Cpp/test/dsa/lib/utils/algorithms/strip_test.cpp
#include <array>
#include <vector>

#include <dsa/lib/utils/algorithm.hpp>

namespace dsa {
namespace {

template<class RandomAccessIterator>
void assertLStrip(RandomAccessIterator expected, RandomAccessIterator first, RandomAccessIterator last) {
    EXPECT_EQ(expected, lstrip(first, last));
}
TEST(LStripTest, empty_returnFirst) {
    std::vector<int> v = {};
    assertLStrip(v.begin(), v.begin(), v.end());
}
TEST(LStripTest, all0s_returnLast) {
    std::vector<int> v = {0};
    assertLStrip(v.end(), v.begin(), v.end());
}
TEST(LStripTest, all0s_returnLast1) {
    std::vector<int> v = {0, 0, 0};
    assertLStrip(v.end(), v.begin(), v.end());
}
TEST(LStripTest, leftNot0s) {
    std::vector<int> v = {1, 2, 3};
    assertLStrip(v.begin(), v.begin(), v.end());
}
TEST(LStripTest, left0s) {
    std::vector<int> v = {0, 1, 2};
    assertLStrip(v.begin() + 1, v.begin(), v.end());
}
TEST(LStripTest, left0s1) {
    std::vector<int> v = {0, 0, 1, 0, 2};
    assertLStrip(v.begin() + 2, v.begin(), v.end());
}
TEST(LStripTest, constexpr) {
    static constexpr std::array<int, 3> a = {0, 1, 2};
    constexpr auto first = lstrip(a.begin(), a.end());
    static_assert(first == a.begin() + 1);
}

template<class RandomAccessIterator>
void assertRStrip(RandomAccessIterator expected, RandomAccessIterator first, RandomAccessIterator last) {
    EXPECT_EQ(expected, rstrip(first, last));
}
TEST(RStripTest, empty_returnFirst) {
    std::vector<int> v = {};
    assertRStrip(v.begin(), v.begin(), v.end());
}
TEST(RStripTest, all0s_returnFirst) {
    std::vector<int> v = {0};
    assertRStrip(v.begin(), v.begin(), v.end());
}
TEST(RStripTest, all0s_returnFirst1) {
    std::vector<int> v = {0, 0, 0};
    assertRStrip(v.begin(), v.begin(), v.end());
}
TEST(RStripTest, rightNot0s) {
    std::vector<int> v = {1, 2, 3};
    assertRStrip(v.end(), v.begin(), v.end());
}
TEST(RStripTest, right0s) {
    std::vector<int> v = {1, 2, 0};
    assertRStrip(v.end() - 1, v.begin(), v.end());
}
TEST(RStripTest, right0s1) {
    std::vector<int> v = {1, 0, 2, 0, 0};
    assertRStrip(v.end() - 2, v.begin(), v.end());
}
TEST(RStripTest, constexpr) {
    static constexpr std::array<int, 3> a = {1, 2, 0};
    constexpr auto last = rstrip(a.begin(), a.end());
    static_assert(last == a.end() - 1);
}

template<class RandomAccessIterator>
void assertStrip(RandomAccessIterator expectedFirst,
                 RandomAccessIterator expectedLast,
                 RandomAccessIterator first,
                 RandomAccessIterator last) {
    auto[actualFirst, actualLast] = strip(first, last);
    EXPECT_EQ(expectedFirst, actualFirst);
    EXPECT_EQ(expectedLast, actualLast);
}
TEST(StripTest, empty_returnFirstLast) {
    std::vector<int> v = {};
    assertStrip(v.begin(), v.end(), v.begin(), v.end());
}
TEST(StripTest, all0s_returnLast) {
    std::vector<int> v = {0};
    assertStrip(v.end(), v.end(), v.begin(), v.end());
}
TEST(StripTest, all0s_returnLast1) {
    std::vector<int> v = {0, 0, 0};
    assertStrip(v.end(), v.end(), v.begin(), v.end());
}
TEST(StripTest, leftRightNot0s) {
    std::vector<int> v = {1, 2, 3};
    assertStrip(v.begin(), v.end(), v.begin(), v.end());
}
TEST(StripTest, leftRight0s) {
    std::vector<int> v = {0, 1, 2, 0};
    assertStrip(v.begin() + 1, v.end() - 1, v.begin(), v.end());
}
TEST(StripTest, leftRight0s1) {
    std::vector<int> v = {0, 0, 1, 0, 2, 0, 0};
    assertStrip(v.begin() + 2, v.end() - 2, v.begin(), v.end());
}
TEST(StripTest, constexpr) {
    static constexpr std::array<int, 4> a = {0, 1, 2, 0};
    constexpr auto actual = strip(a.begin(), a.end());
    static_assert(actual.first == a.begin() + 1);
    static_assert(actual.second == a.end() - 1);
}

}
}
